lua/musicutil.lua

```diff
@@ -0,0 +1,186 @@
+  {name = "Mixolydian Bebop", intervals = {0, 2, 4, 5, 7, 9, 10, 11, 12}},
+  {name = "Blues Scale", alt_names = {"Blues"}, intervals = {0, 3, 5, 6, 7, 10, 12}},
+  {name = "Diminished Whole Half", intervals = {0, 2, 3, 5, 6, 8, 9, 11, 12}},
+  {name = "Diminished Half Whole", intervals = {0, 1, 3, 4, 6, 7, 9, 10, 12}},
+  {name = "Neapolitan Major", intervals = {0, 1, 3, 5, 7, 9, 11, 12}},
+  {name = "Hungarian Major", intervals = {0, 3, 4, 6, 7, 9, 10, 12}},
+  {name = "Harmonic Major", intervals = {0, 2, 4, 5, 7, 8, 11, 12}},
+  {name = "Hungarian Minor", intervals = {0, 2, 3, 6, 7, 8, 11, 12}},
+  {name = "Lydian Minor", intervals = {0, 2, 4, 6, 7, 8, 10, 12}},
+  {name = "Neapolitan Minor", intervals = {0, 1, 3, 5, 7, 8, 11, 12}},
+  {name = "Major Locrian", intervals = {0, 2, 4, 5, 6, 8, 10, 12}},
+  {name = "Leading Whole Tone", intervals = {0, 2, 4, 6, 8, 10, 11, 12}},
+  {name = "Six Tone Symmetrical", intervals = {0, 1, 4, 5, 8, 9, 11, 12}},
+  {name = "Arabian", intervals = {0, 2, 4, 5, 6, 8, 10, 12}},
+  {name = "Balinese", intervals = {0, 1, 3, 7, 8, 12}},
+  {name = "Byzantine", intervals = {0, 1, 3, 5, 7, 8, 11, 12}},
+  {name = "Hungarian Gypsy", intervals = {0, 2, 4, 6, 7, 8, 10, 12}},
+  {name = "Persian", intervals = {0, 1, 4, 5, 6, 8, 11, 12}},
+  {name = "East Indian Purvi", intervals = {0, 1, 4, 6, 7, 8, 11, 12}},
+  {name = "Oriental", intervals = {0, 1, 4, 5, 6, 9, 10, 12}},
+  {name = "Double Harmonic", intervals = {0, 1, 4, 5, 7, 8, 11, 12}},
+  {name = "Enigmatic", intervals = {0, 1, 4, 6, 8, 10, 11, 12}},
+  {name = "Overtone", intervals = {0, 2, 4, 6, 7, 9, 10, 12}},
+  {name = "Eight Tone Spanish", intervals = {0, 1, 3, 4, 5, 6, 8, 10, 12}},
+  {name = "Prometheus", intervals = {0, 2, 4, 6, 9, 10, 12}},
+  {name = "Gagaku Rittsu Sen Pou", intervals = {0, 2, 5, 7, 9, 10, 12}},
+  {name = "Gagaku Ryo Sen Pou", intervals = {0, 2, 4, 7, 9, 12}},
+  {name = "Zokugaku Yo Sen Pou", intervals = {0, 3, 5, 7, 10, 12}},
+  {name = "In Sen Pou", intervals = {0, 1, 5, 2, 8, 12}},
+  {name = "Okinawa", intervals = {0, 4, 5, 7, 11, 12}},
+  {name = "Chromatic", intervals = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12}}
+}
+-- Scale data from https://github.com/fredericcormier/WesternMusicElements
+
+
+--- Generate scale from a root note.
+-- @param root_num MIDI note number (0-127) where scale will begin
+-- @param scale_type String defining scale type (eg, "major", "aeolian" or "neapolitan major"), see class for full list
+-- @param[opt] octaves Number of octaves to return, defaults to 1
+-- @return Array of MIDI note numbers
+function MusicUtil.generate_scale(root_num, scale_type, octaves)
+  if type(root_num) ~= "number" or root_num < 0 or root_num > 127 then return nil end
+  scale_type = scale_type or 1
+  octaves = octaves or 1
+
+  -- Lookup by name
+  if type(scale_type) == "string" then
+    scale_type = string.lower(scale_type)
+    for i = 1, #MusicUtil.SCALES do
```

Comment:
i'm not positive but at a quick glance it seems you could make this a little simpler using the `pairs` function.  something like:

```
for _,scale in pairs(MusicUtil.SCALES) do 
  if string.lower(scale.name) == scale_type then
    ...
```

Reply:
(not at an interpreter now so can't try it, forgive me if i'm only muddying the waters!)

Reply:
From [what I've read](https://stackoverflow.com/questions/8955085/should-i-use-ipairs-or-a-for-loop) there is a performance hit for using `pairs`or `ipairs` so I've been tending to avoid them when there's a sequential index. Open to switching though.

Reply:
ah! interesting. would be fun to do some benchmarking. doesn’t hurt to play it safe though.

Reply:
if i understand correctly, `#` operator has the same performance hit (requires iterating over the whole array to determine length, goes til it hits `nil`, is O(n) on table size)

Reply:
Looking at this again I think there are some small changes I should make to avoid using `#` multiple times on the same table in a single function. Will do those changes and stick with `#` rather than `pairs`.